### DISCUSSION TITLE
delay

### DISCUSSION THREAD

**Discussion by DanSkareda** [SpenceKonde/megaTinyCore#371] [Library discussion]
Hi. I try this sketch on attiny414 (this is only part of my program):

void loop()
{
  lcd.setCursor(0, 0);
  adr=adr+1;
  lcd.print(adr);
  delay(500);
}

When I compile this sketch with last version MegaTinyCore (I instaled it to "hardware" program directory manually), delay not works. 
When I compile this sketch with some older version MegaTinyCore (I installed it through "json" web adress), all ok, delay works.
Is it fault, or I have something wrong set up ? Thanks.

**Comment by SpenceKonde** [comment]
Thanks, I'll see what I managed to break with delay - I did make some changes around delay a couple of days ago that are in master, but not in released version, I guess I didn't actually test whether the new delay implementation gave the right delay >.> 

I had realized the old delay implementation would break when a very long delay (more than 4.2 million milliseconds - ie, someone trying to wait for a couple of hours with delay - which is bad practice, but, you know people will do it). further investigation indicated that it was done to save 24 bytes of flash - but if the new implementation just straight up doesn't work, well... there is more to it I suppose...  On a 4k part you should be getting the new implementation (I kept the old one for 2k parts where that 24 bytes really matters), but if its not behaving like delay should, I need to correct that.

**Reply by SpenceKonde** [reply]
Okay it is linked to some problem with millis timekeeping using TCD0 - switch millis/micros to TCA0, TCB1 or TCB0 for now.... what the hell went wrong with TCD0 now? I didn't change anything related to that!

**Reply by SpenceKonde** [reply]
errrr.... Okay, it fails with TCA0 as well.....

**Reply by SpenceKonde** [reply]
only the TCBs are okay, This is pretty fucking spooky.....

**Reply by SpenceKonde** [reply]
It's what I was most afraid of.... we have backwards time travel, and did before this change too...

Worse still, it has been going on for a while (2.2.9 micros is also  busted - delay works because the stupid wrong implementation of delay() there does not implicitly assume the arrow of time (last year at about this time I spent like a month getting all the backward time travel bugs out of the megaTinyCore timingcodebase. 

Now they're back. Not just one either, TWO, because it's hosed with a type A or D timer... and I do not know how long it has been like this.... fuck me with a flaming pitchfork....

```
Maximum increment 17
Minimum increment 6
Average increment 7.54
Maximum increment 17
Minimum increment 6
Average increment 7.55
Maximum increment 17
Minimum increment 6
Average increment 7.57
Maximum increment 16
Minimum increment 6
Average increment 7.57
Maximum increment 17
Minimum increment 6
Average increment 7.55
Maximum increment 15
Minimum increment 6
Average increment 7.55
Maximum increment 17
Minimum increment 6
Average increment 7.57
Maximum increment 17
Minimum increment 6
Average increment 7.57
Maximum increment 826
Minimum increment -803
Average increment 7.54
Time travel detected 1 times!
```

Also happens infrequently enough that my old test sketch usually misses it :-/

A single instance of backwards time travel, of course, will shortcircuit any delay.

**Reply by SpenceKonde** [reply]
Broken in 2.0.0,. worked in 1.1.10..... no changes that could have conceivably had any effect on this were made between the 1.1.10 and 2.0.0

**Reply by DanSkareda** [reply]
I don't know if I understand what was written well, but I was looking for a version of MegaTinyCore on my PC. The version installed using the "http://drazzy.com/package_drazzy.com_index.json" link is probably 2.2.9, it works properly (I found it in the location "c: \ Users \ Dan \ AppData \ Local \ Arduino15 \ packages \ megaTinyCore \ hardware \ megaavr \ 2.2.9"). The version downloaded manually is according to the file ChangeLog.md - 2.3.0 (I copied it to "c: \ Users \ Dan \ Documents \ Arduino \ hardware \ megaTinyCore-master"), it does not work properly - the delay is always zero, with the parameter change the delay does not change.

**Reply by SpenceKonde** [reply]
Yeah.... I figured out the problem - A couple of days ago to fix an overflow bug in delay that caused delay(4300000) to delay for only 5 seconds I switched to the "standard" delay()
The "standard" delay reacts very badly if micros() ever goes backwards. But micros() never should go backwards and I thought that was fixed since 1.1.9.  But... no, my fixes from back then never actually worked for TCA0 and TCD0 as millis source - only TCB as millis worked with9ut micros() glitching, 

I've checked in fixes With TCA0 for millis, all now works as it should, and if TCD0 is used for millis, for now it will use the old delay() implementation, though I need to fix this problem with TCD0 as millis source...

Thanks for reporting this - there were several problems contributing to this....

**Reply by SpenceKonde** [reply]
aaaahahaaa!!! I think I've got it fixed for TCD0 now too! Tests will take mins to finish before I know for sure  (my test sketch now calls micros() 20.000.000 times; 1,000,000 each with 20 slightly different loops that run at different speeds

**Reply by DanSkareda** [reply]
Great! I tried the latest version of the library and the delay is now right. Thanks.
